test/unit/messaging/schemas/remaining-amount.test.js

```diff
@@ -0,0 +1,274 @@
+const REMAINING_AMOUNT = require('../../../mocks/components/remaining-amount')
+
+const schema = require('../../../../app/messaging/schemas/remaining-amount')
+
+let remainingAmount
+
+describe('remaining amount schema', () => {
+  describe('when remaining amount is present and valid', () => {
```

Comment:
Remaining amount could be negative or zero, All we need is that  remaining amount must be number.

Reply:
By the time we get to recovery, we would be dealing with negative remaining amount.

Reply:
and null.

Yes, but this is checking the schema validation

Reply:
We do not want null, that is why null is covered in test.

Reply:
test('validates fail if null remainingAmount', () => {
      const result = schema.validate(null)
      expect(result.error).toBeDefined()
    })

Reply:
I would move this out into its own describe as you have done for the others and add in the other erroneous types in their own describe as mentioned. 

Could use the `describe.each` to limit code duplication for similar types (depending on what Joi validation error they hit, i.e. group all `number.base` together which would be `null`, `"a"`, etc)

Reply:
Also test on integer number too

Reply:
Integer number restriction has been removed, it was added erroneously ==>remainingAmount could be decimal. Check latest merge comment.

Reply:
Yes, but we're testing test cases for the `remainingAmount`.

Because it is required we need to test:
1. undefined
2. null
3. value

Because it is a float number we need to test:
1. expected float value
2. negative float value
3. integer value
4. stringed expected float value
5. stringed negative float value
6. stringed integer value
7. "a"

We always test for other value types:
1. true
2. false
3. {}
4. []

Reply:
`when remaining amount is **present and valid**` is very implicit whereas each of the other `describe` blocks is explicit `when remaining amount is "a"`

Also, as mentioned the tests covered in `when remaining amount is present and valid` are covered elsewhere in the test with the exception of  when `null` and `undefined` which  is done better in other test files: https://github.com/DEFRA/ffc-pay-statement-generator/blob/187054_Reduction-Schedule-content-update/test/unit/messaging/schemas/document-reference.test.js#L60https://github.com/DEFRA/ffc-pay-statement-generator/blob/187054_Reduction-Schedule-content-update/test/unit/messaging/schemas/document-reference.test.js#L60 &  https://github.com/DEFRA/ffc-pay-statement-generator/blob/187054_Reduction-Schedule-content-update/test/unit/messaging/schemas/document-reference.test.js#L91

Reply:
We are not using any, hence we do not need to check for general values. no need to create redundant test.

Reply:
test('validates fail if undefined remainingAmount', () => {
      const result = schema.validate(undefined)
      expect(result.error).toBeDefined()
    })
    
This test case (when `undefined`) is incomplete.

Simply checking that `result.error` is defined is not enough.

As you have for `when remaining amount is "a"` we need to check for:
1. returns an object
2. returns an object with 2 keys
3. returns an object with "value" key
4. returns remainingAmount for key "value"
5. returns an object with "error" key
6. returns an Error for key "error"
7. returns an object with 4 keys for key "error.details[0]"
8. returns an object with "type" key for key "error.details[0]"
9. returns "<whatever Joi's type for this issue is>" for key "error.details[0].type"
10. returns an object with "message" key for key "error.details[0]"
11. returns "<suitable error message upto yourself>"  for key "error.details[0].message"


I will not pass this PR until these tests are in-place for the above scenarios given it is a simple copy-paste job from `test/unit/messaging/schemas/document-reference.test.js`

Reply:
Reason this seems so long, drawn out and large is becuase this part of adding the `remainingAmount` field and validating in should have been 2 separate tickets each w/ 3 story points.

Because it's all been cramped into 1 means the quality has dropped, the change is large and God knows how Anwar could concrete on testing this ticket as it involves so many different scenarios.

+ Add message item: https://dev.azure.com/defragovuk/DEFRA-FFC/_workitems/edit/183787
+ Validate message item: https://dev.azure.com/defragovuk/DEFRA-FFC/_workitems/edit/191413

Reply:
I am disappointed with your comment, did you manage to check the file at all?
https://github.com/DEFRA/ffc-pay-statement-generator/blob/187054_Reduction-Schedule-content-update/test/unit/messaging/schemas/remaining-amount.test.js

Reply:
Asyncapi added

Reply:
> I am disappointed with your comment, did you manage to check the file at all? https://github.com/DEFRA/ffc-pay-statement-generator/blob/187054_Reduction-Schedule-content-update/test/unit/messaging/schemas/remaining-amount.test.js


Yes, it's added for the test case when it is "a" but not when it is null, undefined, true, false and the integer/negative number cases

Reply:
Test review ticket added.

Reply:
[Review joi component test in statement-generator](https://dev.azure.com/defragovuk/DEFRA-FFC/_workitems/edit/200026)